zkp: Musig2 module

GeneFerneau opened the pull request:
Adds high-level Rust bindings for Musig2 multi-signature protocol

GeneFerneau committed:
secp256k1-zkp-sys: vendor to jonasnick/secp256k1-zkp#musig2

GeneFerneau committed:
secp256k1-zkp-sys: Add Rust FFI for Musig2 module

GeneFerneau committed:
zkp: Add Musig2 module

GeneFerneau commented:
Diff looks **much** larger than it actually is, since I pulled in @jonasnick's changes to `secp256k1-zkp`.

Main changes are in [secp256k1-zkp-sys/src/zkp.rs](https://github.com/GeneFerneau/rust-secp256k1-zkp-1/blob/musig2/secp256k1-zkp-sys/src/zkp.rs) and [rust-secp256k1-zkp/src/zkp/musig.rs](https://github.com/GeneFerneau/rust-secp256k1-zkp-1/blob/musig2/src/zkp/musig.rs).

Will rebase, and update as changes are merged from [#131](https://github.com/ElementsProject/secp256k1-zkp/pull/131)

GeneFerneau commented:
Updated on [#131/0c43aae](https://github.com/jonasnick/secp256k1-zkp/pull/131/commits/0c43aae082368084fc5ec5b3eeea4e4de232ea8f)

elsirion commented on src/zkp/musig.rs:
```diff
@@ -0,0 +1,1158 @@
+        let cx = *secp.ctx();
+        let xonlys = pubkeys
+            .iter()
+            .map(|k| ffi::xonly_from_pubkey(cx, k.as_ptr()).0)
```
Is there a particular reason for not directly using schnorr public keys which already are XOnly internally?

GeneFerneau commented:
Rebased on latest master [ab7f6d8](https://github.com/ElementsProject/rust-secp256k1-zkp/commit/ab7f6d8f5e74fd7bf5348d4fc3593bcc1dd532ad)

Updated to use `schnorrsig::PublicKey`.

Still need to rebase on @jonasnick's latest work, but may wait until the work is finished in their PR.

elsirion commented on src/zkp/musig.rs:
```diff
@@ -0,0 +1,1172 @@
+        &self,
+        secp: &Secp256k1<C>,
+        partial_sigs: &[MusigPartialSignature],
+    ) -> Result<Signature, Error> {
```
I think this `Signature` is still ECDSA?

sanket1729 commented:
Hi @GeneFerneau, musig2 spec and all PRs(but the tweaking support) have been merged upstream. Do you have any plans to resume this?

sanket1729 commented:
Hi @GeneFerneau, I could not resist playing with this branch. I made a few fixups to make it work after #46. Your previous work was very clean, this only required a few changes: https://github.com/sanket1729/rust-secp256k1-zkp/tree/pr29 

If you are busy with other work, I can pick this up
